allennlp/training/optimizers.py

```diff
@@ -26,13 +27,50 @@ class Optimizer(Registrable):
     default_implementation = "adam"
 
     @classmethod
-    def from_params(cls, model_parameters: List[torch.nn.Parameter], params: Params):
+    def from_params(cls, model_parameters: List, params: Params):
         if isinstance(params, str):
             optimizer = params
             params = Params({})
         else:
             optimizer = params.pop_choice("type", Optimizer.list_available())
-        return Optimizer.by_name(optimizer)(model_parameters, **params.as_dict()) # type: ignore
+
+        # make the parameter groups if need
+        groups = params.pop("parameter_groups", None)
+        if groups:
+            # input to optimizer is list of dict
+            # each dict contains {'params': [list of parameters], 'lr': 1e-3, ...}
+            # see: http://pytorch.org/docs/0.3.0/optim.html?#per-parameter-options
+            #
+            # groups contains something like:
+            #"parameter_groups": [
+            #       [['regex1', 'regex2'], {'lr': 1e-3},
+            #        ['regex3'], {'lr': 1e-4}]
+            #]
+            # last entry is for the parameters not in any regex
+            parameter_groups: Any = [{'params': []} for _ in range(len(groups) + 1)]
+            # add the group specific kwargs
+            for k in range(len(groups)): # pylint: disable=consider-using-enumerate
+                parameter_groups[k].update(groups[k][1].as_dict())
+
+            for name, param in model_parameters:
+                # Determine the group for this parameter.
+                group_index = None
+                for k, group_regexes in enumerate(groups):
+                    for regex in group_regexes[0]:
+                        if re.search(regex, name):
```

Comment:
Can you replicate the type of logging that we have for parameter regexes for regularisation? It's very helpful, particularly when you change your model, because you can just see what's changed. 

https://github.com/allenai/allennlp/blob/master/allennlp/nn/initializers.py#L223

Particularly, I would like it if each parameter logged which parameter group it was in, and warnings were raised if you pass parameter regexes which do not match any parameters, because this probably means there was a mis-specified config.